### ISSUE TITLE
Case sensitive `B` in file size literals

### ISSUE THREAD

**Issue by double-daruma** [nushell/nushell#6807]
### Related problem

I was rather surprised to find out that in nushell `20GB == 20Gb` as the GB notation normally stands for Giga**byte** while Gb stands for Giga**bit**. 
nushell treats the `B` or `b` in a file size literal case insensitively.

### Describe the solution you'd like

Since file sizes are measured in bytes and the added complexity of dealing with fractional bytes (bits) does not add much to nushell I propose that file size literals always have to use an uppercase B.

I don't think the SI prefixes (G,k,M,...) need to be case sensitive as nobody is confusing mB for "millibyte"

I believe this breaking change is worth it to avoid user confusion in the future. I'd imagine many faulty scripts would be written under the wrong assumption that "kib" is a kibibit in nushell.

### Describe alternatives you've considered

1. Leave things as they are for compatibility reason.
2. Leave things as they are and implement some sort of lint to point out that Gb should be written as GB instead.
3. Make `8Gib == 1 GiB` and so on: That is add support for sizes in bits to nushell.

**Comment by webbedspace** [comment]
This looks like a disaster waiting to happen.

I support moving all the lowercase bit suffixes from `b` to `bit`, `Gb` -> `gbit`.

**Comment by double-daruma** [comment]
@webbedspace I'm not aware of any file system/file writing API that would allow you to write something smaller than a byte. So, as I have mentioned in my original feature request, I'm not sure if nushell has any need for the File Size type to express bits. It might be significantly easier to keep the File Size *type* itself the way it is (only able to express bytes) while changing the File Size literal *syntax* to avoid confusion.

**Comment by webbedspace** [comment]
Does anyone else have an opinion on this? I'm starting to wonder how many people are also making this mistake…

**Comment by KAAtheWiseGit** [comment]
This was fixed at some point and now `GB`, `Gb`, `gB`, and `gb` all mean the same thing -- a gigabyte.

**Comment by amtoine** [comment]
good catch, this appears to have been fixed, thanks @KAAtheWiseGit :pray:

**Comment by double-daruma** [comment]
No I don't think this has been fixed.

> GB, Gb, gB, and gb all mean the same thing -- a gigabyte.

This is the same behavior as it was when this issue was created. 

I think that this is an issue and a footgun. I am of the opinion that the `b` in a file size literal should be treated in a case sensitive way. The current behavior results in `20GB == 20Gb` which I personally think is confusing.

**Comment by double-daruma** [comment]
@amtoine I think this issue might need to be reopened. Unless the nushell team has decided that the behavior as I described it in the original issue description is the intended behavior. To repeat: In nushell 0.86.0 the expression `20GB == 20Gb` still evaluates to true.

I am not calling for this issue to be reopened because I would oppose the nushell team deciding that this is intended behavior. I am asking for it to be reopened because I believe it was closed due to a misunderstanding.

Thank you for your work on the future of shells!

**Comment by KAAtheWiseGit** [comment]
Ah, that's right, I made a mistake @amtoine.

I guess this would need to be discussed during a team meeting to get a quorum?  I think the issue might've gone unnoticed by the core team the first time it was filed.

**Comment by amtoine** [comment]
no worries :yum: 

we can bring this up in the next meeting for sure if there's a misunderstanding :+1: 


in the meantime, why should `20Gb` and `20GB` not be the same?
or another question: what should `20Gb` and `20GB` mean respectively? :innocent:

**Comment by KAAtheWiseGit** [comment]
The argument @double-daruma makes is that lowercase `b` should stand for bits, while uppercase `B` should be bytes.  So 20Gb would be 2.5GB.  And 20GB is 160Gb.  This [Wikipedia table](https://en.wikipedia.org/wiki/Byte#Multiple-byte_units) has a full description.

**Comment by amtoine** [comment]
> The argument @double-daruma makes is that lowercase `b` should stand for bits, while uppercase `B` should be bytes. So 20Gb would be 2.5GB. And 20GB is 160Gb.
>
oooh wow, i get it now :open_mouth:

**Comment by KAAtheWiseGit** [comment]
One issue I see with it is people not being aware of this.  I wasn't, until doing some further research.  I wonder if it would make sense for forbid all identifiers except for both letters being uppercase?

So in case someone writes `gB` there is a generic error telling them to use `GB`, and if someone uses `Gb`, there is a help message saying that Nushell only supports byte values.

**Comment by double-daruma** [comment]
@KAAtheWiseGit  that overall sounds like a great idea! Note however that the metric kkilobyte "kB" is usually written with a lowercase letter https://en.wikipedia.org/wiki/Kilobyte. For all other sizes including metric and ISO/IEC 80000 the prefix are uppercase.

**Comment by KAAtheWiseGit** [comment]
Related: #8872.

**Comment by double-daruma** [comment]
Sorry I don't mean to annoy you, but if no decision has been made on this yet then maybe this issue should be reopened to make sure it is not forgotten?

**Comment by amtoine** [comment]
ooh yeah for sure, we haven't had a decision on that topic for now :relieved:

**Comment by fdncred** [comment]
Would file sizes still be maintained as bytes? Support bits seems a little odd for file sizes.
https://github.com/nushell/nushell/blob/d717e8faeb5a7e842b47de20fe573ba05a628f26/crates/nu-protocol/src/value/mod.rs#L3746-L3778

Seems like https://github.com/magiclen/byte-unit is the crate we use for file size formatting. 

My guess is this is all lumped in as "FileSize" in SyntaxShape and FlatShape parsing. For people wanting to make a change I'd start here.
https://github.com/nushell/nushell/blob/d717e8faeb5a7e842b47de20fe573ba05a628f26/crates/nu-parser/src/parser.rs#L2216-L2328

And then here.
https://github.com/nushell/nushell/blob/d717e8faeb5a7e842b47de20fe573ba05a628f26/crates/nu-protocol/src/value/unit.rs#L4-L33

**Comment by KAAtheWiseGit** [comment]
@amtoine thanks!

@fdncred I'd honestly just throw an error if someone types lowercase `b`, telling them to use `B` instead and putting a link to somewhere, which elaborates the difference.  If Nushell had warnings, it'd be possible to raise one and still interpret the value as bytes, not bits, but that's a separate topic.
